src/twisted/internet/defer.py

```diff
@@ -631,7 +631,7 @@ def chainDeferred(self, d: "Deferred[_DeferredResultT]") -> "Deferred[None]":
         d._chainedTo = self
         return self.addCallbacks(d.callback, d.errback)
 
-    def callback(self, result: _DeferredResultT) -> None:
+    def callback(self, result: Union[_DeferredResultT, Failure]) -> None:
```

Comment:
I don't understand why we have _DeferredResultT in the first place.

Shouldn't this be `object` as you can pass anything to a callback?

```suggestion
    def callback(self, result: object) -> None:
```

and we have  `def _startRunCallbacks(self, result: object) -> None:`

----------------

Maybe also add a docstring update that `result` can also be a Failure, and in this case this has the same effect as calling `errback`.

Reply:
Type variables are bound to a specific type at some point during type-checking. The scope of that binding is poorly documented (i.e. it's obvious to people who write typecheckers). But in this case it's a type parameter of the Deferred object's type. In the same way that calling `pop()` on a `List[sometype]` returns specifically `sometype` and not `object`, this annotation says you're supposed to call `callback()` with the specific type that it's expecting.

The type annotation on Deferred isn't quite correct anyway (see my comment in [trac 9816](https://twistedmatrix.com/trac/ticket/9816#comment:1) but it may be the best we can do within mypy's limitations.

I do think that if we want to explicitly support the behavior that calling `callback` with a `Failure` is equivalent to calling `errback` then it should be documented.

Reply:
I did think it was odd to pass a `Failure` to `callback`, but that exists in Twisted.  Let me look again as to what code does that, but I don't think we can change the behavior of `callback` without breaking compatibility.

Reply:
Docstring updated.

We don't really have a choice about supporting that behavior; changing it would be a compatibility issue, and probably not a small one.